litmus-portal/graphql-server/pkg/database/mongodb/operations.go

```diff
@@ -44,32 +44,34 @@ func UpdateCluster(query bson.D, update bson.D) error {
 	return nil
 }
 
-func UpsertWorkflowRun(workflow_id string, wfRun WorkflowRun) error {
+func UpsertWorkflowRun(workflow_id string, wfRun WorkflowRun) (int, error) {
 	opts := options.Update().SetUpsert(true)
 	ctx, _ := context.WithTimeout(backgroundContext, 10*time.Second)
 
 	count, err := workflowCollection.CountDocuments(ctx, bson.M{"workflow_id": workflow_id, "workflow_runs.workflow_run_id": wfRun.WorkflowRunID})
 	if err != nil {
-		return err
+		return 0, err
 	}
 
+	updateCount := 1
 	if count == 0 {
 		filter := bson.M{"workflow_id": workflow_id}
 		update := bson.M{"$push": bson.M{"workflow_runs": wfRun}}
 		_, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
 		if err != nil {
-			return err
+			return 0, err
 		}
 	} else if count == 1 {
-		filter := bson.M{"workflow_id": workflow_id, "workflow_runs.workflow_run_id": wfRun.WorkflowRunID}
-		update := bson.M{"$set": bson.M{"workflow_runs.$.last_updated": wfRun.LastUpdated, "workflow_runs.$.execution_data": wfRun.ExecutionData}}
-		_, err = workflowCollection.UpdateOne(ctx, filter, update, opts)
+		filter := bson.M{"workflow_id": workflow_id, "workflow_runs.workflow_run_id": wfRun.WorkflowRunID, "workflow_runs.completed": false}
+		update := bson.M{"$set": bson.M{"workflow_runs.$.last_updated": wfRun.LastUpdated, "workflow_runs.$.execution_data": wfRun.ExecutionData, "workflow_runs.$.completed": wfRun.Completed}}
+		updateResp, err := workflowCollection.UpdateOne(ctx, filter, update, opts)
 		if err != nil {
-			return err
+			return 0, err
 		}
+		updateCount = int(updateResp.MatchedCount)
```

Comment:
Also use `updateResp.Modified` instead  `updateResp.MatchedCount`

Reply:
I followed the documentation for UpdateOne() 
```
If the filter does not match any documents, the operation will succeed and an UpdateResult with a MatchedCount of 0 will be returned.
```

Reply:
I tried in my local system, the query is matched but I am updating a field which is not present.
it returns `MatchedCount: 1, Modified:0 `
```
> db.col.find()
{ "_id" : ObjectId("5fa19715e044d519bc185e97"), "id" : 2 }
> db.col.update({"id": 2}, {"idd": 2})
WriteResult({ "nMatched" : 1, "nUpserted" : 0, "nModified" : 1 })

```

Reply:
We have to add both conditions.

Reply:
this is an impossible situation in our case because our apis and db schemas are strict. Neither can the agent send any data that isn't accounted for in the gql schema nor can it be marshalled to the db struct as it is also strongly typed.

Reply:
We can leave this part. we have to thing more edge case for the test cases.